### PR TITLE
Executing rules inside a function that recovers from panics.

### PR DESCRIPTION
**Pull request by robcapo**
This will prevent a panic inside a rule from impacting the rest of the linter rules.

### CODE REVIEW

File: lint/linter.go [hunk 1]
```diff
@@ -102,3 +103,17 @@ func (l *Linter) run(req Request) (Response, error) {
 
 	return resp, err
 }
+
+func (l *Linter) runAndRecoverFromPanics(rl Rule, req Request) (probs []Problem, err error) {
```
Comment (mingzhi) [inline lint/linter.go]: Can you add a test? Thanks!

Comment (robcapo) [reply lint/linter.go]: Done.

File: lint/config.go [hunk 2]
```diff
@@ -109,6 +109,16 @@ func ReadConfigsYAML(f io.Reader) (Configs, error) {
 	return c, nil
 }
 
+// MakeConfigsForAllPaths takes an input RuleConfig and applies it to all Rules and all file paths.
+func MakeConfigsForAllRulesAndPaths(rc RuleConfig) Configs {
```
Comment (mingzhi) [inline lint/config.go]: Do we need to make it public? Can we move it to linter_test.go?

Comment (robcapo) [reply lint/config.go]: Yeah, we can move it in. I made it public because I figured it would be useful to other tests, but we can move it if we need it.

Comment (mingzhi) [reply lint/config.go]: I think it is better to make it private and only for tests, so we don't need to maintain it.

File: lint/linter_test.go [hunk 3]
```diff
@@ -109,3 +110,29 @@ func TestLinter_run(t *testing.T) {
...
+func (r *panickingRule) Lint(_ Request) ([]Problem, error) { panic("panic") }
+
+func TestLinter_LintProtos_RulePanics(t *testing.T) {
+	r, err := NewRules(&panickingRule{})
+	if err != nil {
+		t.Fatalf("Failed to create Rules: %q", err)
+	}
+
+	fd := new(descriptorpb.FileDescriptorProto)
+	fd.SourceCodeInfo = new(descriptorpb.SourceCodeInfo)
```
Comment (mingzhi) [inline lint/linter_test.go]: Do we need this line?

File: lint/linter_test.go [hunk 4]
```diff
@@ -109,3 +110,29 @@ func TestLinter_run(t *testing.T) {
...
+
+	// linter with only one rule, and a default configuration which enables all rules for all files
```
Comment (mingzhi) [inline lint/linter_test.go]: Remove this comment -- the code is clear enough.

File: lint/linter.go [hunk 5]
```diff
@@ -102,3 +103,17 @@ func (l *Linter) run(req Request) (Response, error) {
...
+	defer func() {
+		if r := recover(); r != nil {
+			if rerr, ok := r.(error); ok {
```
Comment (mingzhi) [inline lint/linter.go]: Can we test this condition?

File: lint/linter_test.go (hunk 3 above)
Comment (robcapo) [reply lint/linter_test.go]: Yes, otherwise creating request will fail.

**Pull request merged**